### PR TITLE
PROTOTYPE: add go.mod support

### PR DESCRIPTION
**Pull request by sttts**
This PR 
- ports the upstream `hack/update-vendor.sh` script to origin
- adds a `hack/bump.sh` script to sync a `glide.yaml` to `go.mod`:

```
$ hack/bump.sh all && hack/update-vendor.sh
```

or just some dependencies:

```
$ hack/bump.sh github.com/openshift/library-go && hack/update-vendor.sh
```

### CODE REVIEW

**Comment by sttts** [comment]
We are not completely there yet, need some more hardcoded dependency versions:
```
# github.com/openshift/origin/vendor/github.com/docker/docker/opts
vendor/github.com/docker/docker/opts/address_pools.go:15:12: undefined: ipamutils.NetworkToSplit
# github.com/openshift/origin/vendor/k8s.io/kube-openapi/pkg/schemaconv
vendor/k8s.io/kube-openapi/pkg/schemaconv/smd.go:104:6: unknown field 'Untyped' in struct literal of type schema.Atom
vendor/k8s.io/kube-openapi/pkg/schemaconv/smd.go:119:14: tr.Inlined.Untyped undefined (type schema.Atom has no field or method Untyped)
vendor/k8s.io/kube-openapi/pkg/schemaconv/smd.go:119:26: undefined: schema.Untyped
vendor/k8s.io/kube-openapi/pkg/schemaconv/smd.go:127:3: a.Struct undefined (type *schema.Atom has no field or method Struct)
vendor/k8s.io/kube-openapi/pkg/schemaconv/smd.go:127:14: undefined: schema.Struct
vendor/k8s.io/kube-openapi/pkg/schemaconv/smd.go:131:4: a.Struct undefined (type *schema.Atom has no field or method Struct)
vendor/k8s.io/kube-openapi/pkg/schemaconv/smd.go:230:5: a.Untyped undefined (type *schema.Atom has no field or method Untyped)
vendor/k8s.io/kube-openapi/pkg/schemaconv/smd.go:230:17: undefined: schema.Untyped
vendor/k8s.io/kube-openapi/pkg/schemaconv/smd.go:232:5: a.Untyped undefined (type *schema.Atom has no field or method Untyped)
vendor/k8s.io/kube-openapi/pkg/schemaconv/smd.go:232:17: undefined: schema.Untyped
vendor/k8s.io/kube-openapi/pkg/schemaconv/smd.go:232:17: too many errors
# github.com/openshift/origin/vendor/k8s.io/apiserver/pkg/server/routes
vendor/k8s.io/apiserver/pkg/server/routes/metrics.go:36:23: undefined: prometheus.Handler
vendor/k8s.io/apiserver/pkg/server/routes/metrics.go:46:27: undefined: prometheus.Handler
# github.com/openshift/origin/vendor/k8s.io/kubernetes/pkg/kubectl/util/templates
vendor/k8s.io/kubernetes/pkg/kubectl/util/templates/markdown.go:30:5: cannot use &ASCIIRenderer literal (type *ASCIIRenderer) as type blackfriday.Renderer in assignment:
	*ASCIIRenderer does not implement blackfriday.Renderer (missing RenderFooter method)
vendor/k8s.io/kubernetes/pkg/kubectl/util/templates/markdown.go:66:11: undefined: blackfriday.LIST_ITEM_BEGINNING_OF_LIST
vendor/k8s.io/kubernetes/pkg/kubectl/util/templates/markdown.go:73:11: undefined: blackfriday.LIST_TYPE_ORDERED
vendor/k8s.io/kubernetes/pkg/kubectl/util/templates/normalizers.go:73:35: too many arguments to conversion to blackfriday.Markdown: blackfriday.Markdown(bytes, composite literal, blackfriday.EXTENSION_NO_INTRA_EMPHASIS)
246,12s user 55,66s system 844% cpu 35,724s total
```

File: hack/update-vendor.sh [hunk 1]
```diff
@@ -0,0 +1,415 @@
...
+fi
+
+# TODO: kube::golang::verify_go_version
+# TODO: kube::util::require-jq
+
+TMP_DIR="${TMP_DIR:-$(mktemp -d /tmp/update-vendor.XXXX)}"
+LOG_FILE="${LOG_FILE:-${TMP_DIR}/update-vendor.log}"
+kube::log::status "logfile at ${LOG_FILE}"
+
+if [ -z "${BASH_XTRACEFD:-}" ]; then
```
Comment (deads2k) [inline hack/update-vendor.sh]: this, this is why people think you're weird in bash.  What does this do?

File: hack/update-vendor.sh [hunk 2]
```diff
@@ -0,0 +1,350 @@
...
+set -o errexit
+set -o nounset
+set -o pipefail
+
+KUBE_ROOT=$(dirname "${BASH_SOURCE[0]}")/..
+source "${KUBE_ROOT}/hack/lib/init.sh"
+
+# Explicitly opt into go modules, even though we're inside a GOPATH directory
+export GO111MODULE=on
+# Explicitly clear GOFLAGS, since GOFLAGS=-mod=vendor breaks dependency resolution while rebuilding vendor
```
Comment (ironcladlou) [inline hack/update-vendor.sh]: Can you elaborate on this? In my team's projects which use modules with vendor enabled, we have no choice but to  use `GOFLAGS=-mod=vendor` for all our tooling to work, and we've had no issues with rebuilding vendor. So I just wonder whether this is idiosyncratic to origin's build (in which case maybe that should be detailed here) or whether you were alluding to some Go issue of which I was not (but maybe should be) aware.

Comment (sttts) [reply hack/update-vendor.sh]: this is copied from upstream. We might want to clean up that stuff around the actually algorithm.

File: hack/update-vendor.sh (hunk 1 above)
Comment (sttts) [reply hack/update-vendor.sh]: don't know, ask @liggitt ;-)

File: hack/update-vendor.sh [hunk 3]
```diff
@@ -187,134 +234,179 @@ group_replace_directives
...
+rm -rf staging-kube
+mv vendor/k8s.io/kubernetes/staging staging-kube
 go mod vendor >>"${LOG_FILE}" 2>&1
+for repo in $(list_staging_repos staging-kube/src/k8s.io); do
+  cp "staging-kube/src/k8s.io/${repo}/go.mod" "vendor/k8s.io/kubernetes/staging/src/k8s.io/${repo}"
+done
+rm -rf staging-kube
+ln -fs vendor/k8s.io/kubernetes/staging staging-kube
+
+# Phase 7: unprune kubernetes by checking it out completely
```
Comment (tnozicka) [inline hack/update-vendor.sh]: I wasn't looking really closely, but can we just unprune everything?

File: hack/update-vendor.sh (hunk 2 above)
Comment (liggitt) [reply hack/update-vendor.sh]: This particular script's job is to build vendor. It should not use the vendor dir while doing so. 

All other scripts can use -mod=vendor

File: hack/update-vendor.sh (hunk 1 above)
Comment (liggitt) [reply hack/update-vendor.sh]: It dumps all invocations to a file. Was handy for debugging.

File: hack/update-vendor.sh (hunk 3 above)
Comment (sttts) [reply hack/update-vendor.sh]: if you add **all** packages to the kube.go, sure ;-)

**Pull request closed without merging**